packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+                                        absl::StrJoin(info->name.fullName, "::", NameFormatter(ctx)));
+                        }
+                    }
+                    info->importedPackageNames.emplace_back(move(name));
+                }
+            }
+        }
+
+        if (send->fun == core::Names::exportMethods()) {
+            const bool alreadyCalled = exportMethodsLoc.exists();
+            if (alreadyCalled) {
+                if (auto e = ctx.beginError(send->loc, core::errors::Packager::MultipleExportMethodsCalls)) {
+                    e.setHeader("export_methods can only be called once in a package");
+                    e.addErrorLine(core::Loc(ctx.file, exportMethodsLoc), "Previous call to export_methods found here");
+                }
+            } else {
+                exportMethodsLoc = send->loc;
+            }
+            for (auto &arg : send->args) {
+                auto target = verifyConstant(ctx, core::Names::exportMethods(), arg);
+                if (target != nullptr) {
+                    exportedMethods.emplace_back(copyConstantLit(target));
+                    // Transform the constant lit to refer to the target within the mangled package namespace.
+                    arg = prependInternalPackageNameToScope(move(arg));
+                }
+            }
+        }
+
+        return tree;
+    }
+
+    ast::TreePtr preTransformClassDef(core::MutableContext ctx, ast::TreePtr tree) {
+        auto *classDef = ast::cast_tree<ast::ClassDef>(tree);
+        if (classDef->symbol == core::Symbols::root()) {
+            // Ignore top-level <root>
+            return tree;
+        }
+
+        if (classDef->ancestors.size() != 1 || !isReferenceToPackageSpec(ctx, classDef->ancestors[0])) {
+            if (auto e = ctx.beginError(classDef->loc, core::errors::Packager::InvalidPackageDefinition)) {
+                e.setHeader("Expected package definition of form `Foo::Bar < PackageSpec`");
+            }
+        } else if (info == nullptr) {
+            info = make_shared<PackageInfo>();
+            info->name = getPackageName(ctx, classDef->name, true);
+            info->loc = core::Loc(ctx.file, classDef->loc);
+        } else {
+            if (auto e = ctx.beginError(classDef->loc, core::errors::Packager::MultiplePackagesInOneFile)) {
+                e.setHeader("Package files can only declare one package");
+                e.addErrorLine(info->loc, "Previous package declaration found here");
+            }
+        }
+
+        return tree;
+    }
+
+    ast::TreePtr postTransformClassDef(core::MutableContext ctx, ast::TreePtr tree) {
+        auto *classDef = ast::cast_tree<ast::ClassDef>(tree);
+        if (classDef->symbol != core::Symbols::root() || info == nullptr || exportedMethods.empty()) {
+            return tree;
+        }
+
+        // Inject <PackageRegistry>::Name_Package::<PackageMethods> within the <root> class
```

Comment:
Why let people choose the name for their `export_methods` namespace? Why not just always force people to use the name `PackageMethods`? And then it wouldn't be necessary to include their name into this name.

Reply:
> Why let people choose the name for their export_methods namespace?

`<PackageMethods>` is an implementation detail; users don't see it.

Every package can export methods from one or more modules via `export_methods`. `<PackageRegistry>::Name_Package` is the package namespace for package `Name`, and `<PackageRegistry>::Name_Package::<PackageMethods>` is a module that contains all of the exported methods from that package.

This setup makes it really easy to import a package's methods into other packages without needing to enumerate the modules whose methods are exported -- you just extend the `<PackageMethods>` for that module.

Reply:
> Every package can export methods from one or more modules via `export_methods`.

But there's an error above that says you're only allowed to call `export_methods` once per package.


> you just extend the `<PackageMethods>` for that module

And I'm saying that it would be the same. You just extend `PackageMethods` for that module, and people don't get to choose a name to give to `export_methods`.

Reply:
> But there's an error above that says you're only allowed to call export_methods once per package.

`export_methods module1, module2, module3`

> You just extend PackageMethods for that module, and people don't get to choose a name to give to export_methods.

Ah. This is a spec suggestion for @nroman-stripe rather than an implementation suggestion.

Reply:
Do you have a test for `export_methods module1, module2, module3`?

Reply:
I'll make sure I do.

Reply:
I added it to `multiple_export_methods`.